### ISSUE TITLE
FileChooser: Directory separator in the filename entry is passed to the caller without validation.

### ISSUE THREAD

**Issue by mcclurgm** [elementary/files#2016]
### What Happened?

When I enter invalid characters into a filename, the filechooser exits silently and then seems to cause undefined behavior in the app. The behavior varies a bit by app in terms of how the resulting file is handled, but the chooser always closes with no warning, as if the save had succeeded. But no file is saved to the file system, at least on either of the apps I tried (Code and Apostrophe).

### Steps to Reproduce

1. Open a new document in Code
2. Press the `Save As` button and navigate to some directory you'll remember
3. Enter a file name with a slash, like `Stuff/Things.txt` and press `Save` to try to save the file in that directory
4. The dialog will close immediately
5. (This is Code specific, but may show part of the behavior:) The portion of the filename you entered (like `Things.txt` appears in the tab title
6. Open Files to the directory you tried to save the file to. The file will not be there, so it was never saved.

If I repeat the same process in Apostrophe, clicking `Save` multiple times and re-entering the same invalid file name will never save a file, so the app will always put up the file chooser dialog. Code will silently fail to save from there on. This is what I mean by how behavior differs. But I don't think the invalid name should be passed to the apps in the first place.

### Expected Behavior

Some warning should appear that the filename contains invalid characters and suggest the appropriate changes.

### OS Version

6.x (Odin)

### Software Version

Compiled from git

**Comment by jeremypw** [comment]
I agree it seems reasonable that apps should be able to rely on the filechooser portal to return a valid filename (or require the user to cancel the operation).

**Comment by danrabbit** [comment]
Yeah we should probably use a validated entry here and make the save button insensitive if the file name is not valid

**Comment by Marukesu** [comment]
I assume that the `Stuff` folder doesn't exists right? looking at the return value from the SaveFile method, we are returning an valid uri. So, maybe we should create the parent folders if it not exists?

> (This is Code specific, but may show part of the behavior:) The portion of the filename you entered (like Things.txt appears in the tab title

you are using the xdg-desktop-portal from the ubuntu repos right? that don't happen with the latest release. the tab are empty and there's no tooltip too. `save as` are also disabled.

can you run `/usr/libexec/xdg-desktop-portal -vr` and see if it show some warning or if the uri is exported to the portal FUSE? here it's show:

```
WARNING **: 01:45:02.991: Failed to register file:///home/marukesu/Stuff/Thing.txt: Failed to open file:///home/marukesu/Stuff/Thing.txt
```

**Comment by jeremypw** [comment]
In OS8 the filechooser still returns the requested filename `~/Stuff/Things.txt` even though the parent folder "Stuff" does not exist in `~`.  However Code now handles that case properly and throws an error dialog.   I am not sure what the file chooser portal definition requires if the user enters a path or other invalid filename.  It could be argued that the application should be responsible for deciding what to do.  For example, instead of Code rejecting the result, it could have created the parent folder.  The user could also have created the folder within the filechooser UI.  Need to check whether we supply the entry or Gtk.

**Comment by jeremypw** [comment]
Its pretty hard to find an invalid filename that can be entered from the keyboard in Linux to be honest.  `/` is about the only character causing problems.  It would be nice if the filechooser could automatically create parent folder(s) if they don't exist to save going through the dialog.

**Comment by jeremypw** [comment]
@danirabbit Is this something that should be sent upstream so that all filechoosers work the same way?

**Comment by jeremypw** [comment]
Looks like we do provide the entry so we could implement validation and maybe automatic parent creation.  A drawback might be that if the parent is created and the save fails or is cancelled then the parent might not be expected to persist.

**Comment by jeremypw** [comment]
Correction, we do not provide the entry - just modify the Gtk supplied one so we would have to swap it for a Granite.ValidatedEntry.

**Comment by jeremypw** [comment]
Just pushed a PR that does this - seems to work OK.

**Comment by jeremypw** [comment]
Reopening as the merge that fixed this has been reverted as it caused regressions.
